src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -81,8 +81,43 @@ public PrimitiveColumnContainer(IEnumerable<T> values)
 
         public PrimitiveColumnContainer(ReadOnlyMemory<byte> buffer, ReadOnlyMemory<byte> nullBitMap, int length, int nullCount)
         {
-            Buffers.Add(new ReadOnlyDataFrameBuffer<T>(buffer, length));
-            NullBitMapBuffers.Add(new ReadOnlyDataFrameBuffer<byte>(nullBitMap, length));
+            ReadOnlyDataFrameBuffer<T> dataBuffer;
+            if (buffer.IsEmpty)
+            {
+                DataFrameBuffer<T> mutableBuffer = new DataFrameBuffer<T>();
+                mutableBuffer.EnsureCapacity(length);
+                mutableBuffer.RawSpan.Fill(default(T));
+                dataBuffer = mutableBuffer;
+            }
+            else
+                dataBuffer = new ReadOnlyDataFrameBuffer<T>(buffer, length);
+            Buffers.Add(dataBuffer);
+            int bitMapBufferLength = (int)Math.Ceiling(length / 8.0);
+            ReadOnlyDataFrameBuffer<byte> nullDataFrameBuffer;
+            if (nullBitMap.IsEmpty)
+            {
+                if (nullCount != 0)
+                    throw new ArgumentNullException(nameof(nullBitMap));
+                if (!buffer.IsEmpty)
+                {
+                    // Create a new bitMap with all the bits set
+                    var bitMap = new byte[bitMapBufferLength];
+                    for (int i = 0; i < bitMapBufferLength - 1; i++)
+                        bitMap[i] = 255;
+                    for (int i = 0; i < length - (bitMapBufferLength - 1) * 8; i++)
+                    {
+                        bitMap[bitMapBufferLength - 1] = SetBit(bitMap[bitMapBufferLength - 1], i, true);
```

Comment:
(minor) you could probably do this a bit fancier/faster. Since you know how many elements past a multiple of 8 you have, one thought is to take `2^(elementsLeft + 1) - 1`, which will give you a byte with that many bits set. Or another way is described here: [https://stackoverflow.com/a/12416681/3680432](https://stackoverflow.com/a/12416681/3680432).  

> The usual way is to take a 1, and shift it left n bits. That will give you something like: 00100000. Then subtract one from that, which will clear the bit that's set, and set all the less significant bits, so in this case we'd get: 00011111.

Reply:
Ah, so obvious now that I've seen it. Nice!